### ISSUE TITLE
Starting storybook takes very long

### ISSUE THREAD

**Issue by kroeder** [storybooks/storybook#3968]
### Bug or support request summary

I'm not sure if this is just the nature of a big storybook (we got 46 stories so far 😃 ) but starting it takes more and more time (about 2,5 min, currently) 

Also hot-reloading takes about 3-5s (which is "fine" but not smooth compared to the initial experience of an empty project)

I'd like to investigate this further and, if possible, improve this but I don't really know where to start.
One thing I already noticed is, that the build stops at a certain % for quite a while; it takes about 60-70% of the whole build time

First try
```
 33% building modules 193/219 modules 26 active …ress-column\address-column.component.ts
 33% building modules 193/220 modules 27 active …@storybook\ui\dist\modules\ui\routes.js
 33% building modules 194/221 modules 27 active …ok\angular\dist\client\preview\index.js
 33% building modules 195/221 modules 26 active …ok\angular\dist\client\preview\index.js
 33% building modules 195/222 modules 27 active …@storybook\react-komposer\dist\index.js
```

Second try
```
 31% building modules 181/210 modules 29 active …lient\node_modules\es5-shim\es5-shim.js
 32% building modules 184/212 modules 28 active …ybook\addon-storysource\dist\preview.js
 32% building modules 185/212 modules 27 active …ybook\addon-storysource\dist\preview.js
 32% building modules 187/213 modules 26 active …@storybook\ui\dist\modules\ui\routes.js
 32% building modules 189/217 modules 28 active …ient\node_modules\fbjs\lib\invariant.js
```

Third try
```
 31% building modules 182/212 modules 30 active …js\library\modules\es6.object.create.js
 32% building modules 184/213 modules 29 active …odules\json-stringify-safe\stringify.js
 32% building modules 185/213 modules 28 active …odules\json-stringify-safe\stringify.js
 32% building modules 186/213 modules 27 active …odules\json-stringify-safe\stringify.js
 32% building modules 187/214 modules 27 active …@storybook\ui\dist\modules\ui\routes.js
 32% building modules 189/214 modules 25 active …@storybook\ui\dist\modules\ui\routes.js
```

Have you noticed this yourself for larger storybook projects? 
Do you know any tools or options I can use to identify the root of this problem?

We are getting about 10-15 more stories the next couple months I guess. I hope at least the hot reload time does not scale further with the size of our project

### Please specify which version of Storybook and optionally any affected addons that you're running

```
 "@storybook/addon-actions": "^4.0.0-alpha.14",
    "@storybook/addon-knobs": "^4.0.0-alpha.14",
    "@storybook/addon-links": "^4.0.0-alpha.14",
    "@storybook/addon-notes": "^4.0.0-alpha.14",
    "@storybook/addon-options": "^4.0.0-alpha.14",
    "@storybook/addon-storysource": "^4.0.0-alpha.14",
    "@storybook/addons": "^4.0.0-alpha.14",
    "@storybook/angular": "^4.0.0-alpha.14",
```

**Comment by igor-dv** [comment]
You can probably add a [profile](https://webpack.js.org/configuration/other-options/#profile) prop in the custom `.storybook/webpack.config.js`. 

Also, can you share it?

**Comment by kroeder** [comment]
Unfortunately, it's not a public project. I've started to generate some webpack stats files and playing around with different webpack configurations.

## Webpack compiler configuration

Test-Method | Compile Time (in min)
--- | ---
Full-Control-Mode + Default | 4.4, 5.4
Full-Control-Mode + Default + parallelism: 4 | 4.33
Full-Control-Mode + Default + parallelism: 1 | 2.35
--- same test --- | 2.52
--- same test --- | 2.61
--- same test --- | 2.60

So, somehow parallelism runs faster at 1 compared to any other input value (tested 2 as well) 
It's also the most stable one (does not vary between 4.4 and 5.4)

## Story-Loader

I also noticed that the compiler *always* stops right after
```
69% building modules 1298/1299 modules 1 active …t-client\projects sync /\.stories\.ts$/ <----------
 70% building modules 1299/1299 modules 0 active
 69% building modules 1299/1300 modules 1 active …ib\table\table\stories\table.stories.ts
```
Which is triggered by this I guess?
```
const req = require.context('../projects/', true, /\.stories\.ts$/);

function loadStories() {
    req.keys().forEach((filename) => req(filename))
}

configure(loadStories, module);
``` 

Test-Method | Compile Time (in min)
--- | ---
Loading no stories | 0.23
Loading 1 story | 2.03

Loading just 1 story was quite interesting
My config now looks like this 
```
function loadStories() {
    // req.keys().forEach((filename) => req(filename))
    require('../projects/client-ui/src/lib/components/layouts/master-detail/stories/master-detail.stories');
}

configure(loadStories, module);
```
I forced to just require 1 single story but the compiler stucks again right at the first story he tries to pack
```
 69% building modules 1493/1494 modules 1 active …\_esm5\internal\observable\fromEvent.js
 70% building modules 1498/1498 modules 0 active
 70% building modules 1503/1503 modules 0 active
 69% building modules 1503/1504 modules 1 active …stories\master-detail-demo.component.ts
```

It seems it has nothing to do with how many stories I have but with webpack starting to pack the first story. After that it does not stop again.

Is this in any way helpful? I can try to isolate this in a fresh Angular-CLI 6.1 Project with `ng generate library test-lib` and adding a single story to it. Might be tomorrow until I can post the results of this test, though.


## webpack.config.ts
```
module.exports = (baseConfig, env, defaultConfig) => {

    const tsFile = path.resolve(__dirname, '../tsconfig.json'); // it's important that this is absolute for the resolve.alias values!
    const tsConfig = JSON.parse(fs.readFileSync(tsFile).toString('utf8'));
    const root = path.join(path.dirname(tsFile), tsConfig.compilerOptions.baseUrl);
    if (!defaultConfig.resolve.alias) {
        defaultConfig.resolve.alias = {};
    }

    for (const alias in tsConfig.compilerOptions.paths) {
        if (alias.includes('*')) {
            // we have mappings for /* in the tsconfig, but we don't need these in webpack
            continue;
        }
        const targetPath = tsConfig.compilerOptions.paths[alias][0];
        defaultConfig.resolve.alias[alias] = path.join(root, targetPath);
    }

    defaultConfig.parallelism = 1;

    return defaultConfig;
};
```

## config.js
```
import { setOptions } from '@storybook/addon-options';
import { configure, addDecorator  } from '@storybook/angular';
import { withNotes } from '@storybook/addon-notes';

addDecorator(withNotes);

setOptions({
    /**
     * name to display in the top left corner
     * @type {String}
     */
    name: 'Storybook',
    /**
     * URL for name in top left corner to link to
     * @type {String}
     */
    url: '#',
    /**
     * open story component as full screen
     * @type {Boolean}
     */
    goFullScreen: false,
    /**
     * display panel that shows a list of stories
     * @type {Boolean}
     */
    showStoriesPanel: true,
    /**
     * display panel that shows addon configurations
     * @type {Boolean}
     */
    showAddonPanel: true,
    /**
     * display floating search box to search through stories
     * @type {Boolean}
     */
    showSearchBox: false,
    /**
     * open addon panel as a vertical panel on the right
     * @type {Boolean}
     */
    addonPanelInRight: false,
    /**
     * sorts stories
     * @type {Boolean}
     */
    sortStoriesByKind: false,
    /**
     * regex for finding the hierarchy separator
     * @example:
     *   null - turn off hierarchy
     *   /\// - split by `/`
     *   /\./ - split by `.`
     *   /\/|\./ - split by `/` or `.`
     * @type {Regex}
     */
    hierarchySeparator: null,
    /**
     * regex for finding the hierarchy root separator
     * @example:
     *   null - turn off mulitple hierarchy roots
     *   /\|/ - split by `|`
     * @type {Regex}
     */
    hierarchyRootSeparator: null,
    /**
     * sidebar tree animations
     * @type {Boolean}
     */
    sidebarAnimations: true
});

const req = require.context('../projects/', true, /\.stories\.ts$/);

function loadStories() {
    // req.keys().forEach((filename) => req(filename))
    require('../projects/client-ui/src/lib/components/layouts/master-detail/stories/master-detail.stories');
}

configure(loadStories, module);
```

**Comment by igor-dv** [comment]
Very nice research. 
You wrote that you are using `@storybook/addon-storysource`, but I don't see it's configured.

**Comment by kroeder** [comment]
Which tsconfig is used to build storybook? The root tsconfig of a project?

**Comment by igor-dv** [comment]
Storybook itself is not in TS. For angular you have this - https://github.com/storybooks/storybook/blob/master/app/angular/tsconfig.json

**Comment by kroeder** [comment]
Thanks! I found out that I can place a custom tsconfig in the .storybook configuration folder
I'm currently trying to make a minimal reproduction repo but this is really a tough one.

I tried to start with an empty project and add more and more stuff but this does not produce this issue. I'm now going the other way around and remove more and more stuff from our project until the issue is gone. Hopefuly, the result of my work makes it possible to offer you a gitlab repo or at least a way to reproduce this.

Super cool would be some sort of parameter in the CLI to analyze the webpack build `npm run storybook --debug`

I assume using https://github.com/stephencookdev/speed-measure-webpack-plugin/ is only possible in full-control mode without extending the storybook base config? Is this even possible in an angular project with angular-cli?

**Comment by igor-dv** [comment]
> Super cool would be some sort of parameter in the CLI to analyze the webpack build npm run storybook --debug 

Not sure, I got you what it will do

> Is this even possible in an angular project with angular-cli?

Yeah, you can add whatever you want to webpack in the extended/full control of any app

**Comment by andrei-ilyukovich** [comment]
@kroeder @igor-dv I also faced with significant performance issues during starting up storybook. It works fast in case of having one or two simple stories, but adding about 10 leads to start up time about 3-4 mins.
Also I've noticed that addon-knobs for Angular works in a very strange manner that impact story loading performance. Every time I change some knob, the whole story is reinitialized like during story first loading (I assume that based on initializing of modules needed by story, changing value in knob causes modules to be recreated)

**Comment by kroeder** [comment]
Can you reproduce this in a Repo? I was not able to..

**Comment by andrei-ilyukovich** [comment]
Yes, sure. You can open browser console for this page https://storybooks-angular.netlify.com/?selectedKind=Addon%7CKnobs&selectedStory=All%20knobs&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs&background=%2300aced
There should be "constructor" log  message from AllKnobsComponent. Then try to open another story then get back to "All knobs" again - you'll see double "constructor" message. Also changing any value in knob panel forces AllKnobsComponent to be recreated

**Comment by andrei-ilyukovich** [comment]
I didn't find such behavior for stories which doesn't have knobs

**Comment by andrei-ilyukovich** [comment]
Probably performance issue is related somehow to typescript files processing. https://storybook.js.org/configurations/typescript-config/
Guide tells that we need to add "awesome-typescript-loader" to work with typescript, but angular-cli example doesn't have it configured. So do we need it or not? If not then what loader is used by storybook under the hood to process ts files?

**Comment by igor-dv** [comment]
If awesome-typescript-loader solves the performance, we can make it work, Storybook doesn't use angular-cli's `ts` rules.

**Comment by andrei-ilyukovich** [comment]
So could you please explain how [angular-cli example](https://github.com/storybooks/storybook/tree/master/examples/angular-cli) is working without any typescript configuration? Does default webpack config contain typescript loader definition?

**Comment by kroeder** [comment]
I did some more research regarding my initial issue

I said "Loading 1 story takes as long as loading every story (about 40-45)"
That's not entirely correct.

#### Loading 1 story in a project with 40-45 story takes as long as loading every of these 40-45 stories
*Start time: 1,5min
Hot reload time: 5.2s*
```
function loadStories() {
    // req.keys().forEach((filename) => req(filename))
    require("../projects/client-ui/src/lib/components/buttons/button-primary/stories/button-primary.stories")
}

configure(loadStories, module);
```

#### Removing every component + story except 1
*Start time: 0,5min
Hot reload time: 1s*

I prepared a repo with the *1 story* scenario but I guess this is of no interest if it runs so much faster than one with many components, which I can't make public..

In my next steps I'll play around with a seperat tsconfig file for storybook. It seems that starting storybook includes every file that is specified in the root tsconfig.json. I noticed that by removing all of the components. Compile errors occurred in non-story relevant code

**Comment by igor-dv** [comment]
@kroeder, it makes sense the problem is somewhere in ts configuration. Maybe you can just copy-past an existing component 100 times to simulate the problem ?

@andrei-ilyukovich with the full control mode of the webpack.config, you can do something like this (pseudo):

```js
module.exports = (basicConfig, mode, defaultConfig) => {
   basicConfig.module.rule = removeTsRule(basicConfig.module.rule);

   basicConfig.module.rule.push(createAwesomeTsRule())
   
   return basicConfig;
};
```

**Comment by andrei-ilyukovich** [comment]
@kroeder @igor-dv It look like I found solution to reduce building time significantly. The reason why it takes so long in current config is just ts-loader configuration, which is used internally by storybook. We need to skip type checking with setting flag ``` transpileOnly: true``` for ts-loader in webpack config. [Ts-loader docs](https://github.com/TypeStrong/ts-loader) tell us how it is possible to still have faster type checking with "fork-ts-checker-webpack-plugin". These configuration updates reduces building time from 4 mins to 11 sec.  Here is how webpack.config should look like:
```typescript
const path = require("path");
const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");

module.exports = (baseConfig, env, config) => {
// Find rule for processing ts files and then add transpileOnly option to skip type checking
    let rule = config.module.rules.find((rule) => rule.use[0].loader.includes("ts-loader"));
    rule.use[0].options.transpileOnly = true;
// Add plugin to have separate type checking
    config.plugins.push(
        new ForkTsCheckerWebpackPlugin({
            tslint: path.resolve(__dirname, "../tslint.json"),
            tsconfig: path.resolve(__dirname, "tsconfig.json")
        })
    );
    return config;
};
```

Another option is to replace ts-loader with awesome-typescript-loader. With such config build time is about 25 sec, which is good but worse than first solution.
```typescript
const path = require("path");

module.exports = (baseConfig, env, config) => {
// Find rule for processing ts files and then set new loader
    let rule = config.module.rules.find((rule) => rule.use[0].loader.includes("ts-loader"));
    rule.use[0].loader = require.resolve("awesome-typescript-loader");
// "awesome-typescript-loader" has a bit different option name to set tsconfig.json (ts-loader has configFile option)
    rule.use[0].options.configFileName = path.resolve(__dirname, "tsconfig.json");

    return config;
};
```

**Comment by igor-dv** [comment]
😱 omg, it's a huge difference. We probably need to enable `transpileOnly` by default for angular... 
I don't know though if we want to have a ForkTsCheckerWebpackPlugin set by default. It means we will also need to discover the user's `tslint.json`. 

BTW, there is an option to define a custom tsconfig.json in the .storybook dir.

**Comment by andrei-ilyukovich** [comment]
Basically tslint is just an option, you could configure ForkTsCheckerWebpackPlugin to not make ts linting.
But anyway you could use awesome-typescript-loader which also have fast type checking out-of-the box.

Yes, I know about custom tsconfig.json, in examples above I have to set it explicitly, because by default it isn't picked up by plugin

**Comment by kroeder** [comment]
Can confirm.
Using `ForkTsCheckerWebpackPlugin` and `defaultConfig.parallelism = 1;` it took 
*Build Time: 50s~*

Without `defaultConfig.parallelism = 1;` (which was only used, because ts-loader ran faster with it)
*Build Time: 38s~*

Might be quicker but I have do deal with a couple new typescript errors like
```
typings.d.ts(7,5):
TS1038: A 'declare' modifier cannot be used in an already ambient context.
```

or

```@types/storybook__addon-notes/index.d.ts(9,32):
TS2307: Cannot find module '@storybook/react'.```

**Comment by kroeder** [comment]
After fixing everything it takes around 30-35s - Sadly, my hot-reload still takes around 5s but that's ok for now

More good news: It seems this approach moves the build-time bottleneck to something else.
30-35s build time on a Citrix VM
25s~ build time on a Gaming PC

Before:
2:30min build time on a Citrix VM
1:30min build time on a Local Developer Machine (which is obviously weaker than my gaming PC 😂 )

Sorry for throwing around with numbers the whole time but I guess build time is exactly about that 😃 

So: @igor-dv is this something you would consider add to the `@storybook/angular` `defaultConfig` or should we use a custom webpack config? Regarding the `tslint.json` discovery.. In most cases it should be at the same place as the `angular.json`. If not you could add an example of how to configure a custom `tsconfig.json` and custom `tslint.json` location

Maybe too naive?

**Comment by igor-dv** [comment]
I think we need to include `transpileOnly` to the default config, 

Would somebody like to PR this?

**Comment by kroeder** [comment]
`transpileOnly` sounds good but there should be any form of type checking per default
Just setting it to `true` feels wrong to me

What about your concerns regarding `tslint.json`? See my comment above

> Regarding the tslint.json discovery.. In most cases it should be at the same place as the angular.json. If not you could add an example of how to configure a custom tsconfig.json and custom tslint.json location

**Comment by igor-dv** [comment]
Makes sense

**Comment by NiklasWebnuts** [comment]
@nicholasmaddren Did you find any performance optimisations?

**Comment by kroeder** [comment]
@igor-dv @ndelangen do you know anything about that issue?

**Comment by fonkgoku** [comment]
@kroeder I had the same problem. Removing `react-docgen-typescript-loader` from `./storybook/webpack.config.js` improved my performance ~5-fold (both on build and hot-reload). If you can live without having your typescript props displayed in your addons section (see the description: https://github.com/strothj/react-docgen-typescript-loader) this seems like a quick an easy fix.

**Comment by RoystonS** [comment]
If you haven't already, it's worth ensuring you've provided an explicit `tsconfig.json` file to both your loaders.

e.g. for the docgen loader:

```

**Comment by RoystonS** [comment]
If you haven't already, it's worth ensuring you've provided an explicit `tsconfig.json` file to your loaders, especially the docgen loader, e.g.

```
      {
        loader: require.resolve('react-docgen-typescript-loader'),
        options: {
          tsconfigPath: path.resolve(__dirname, '../stories/tsconfig.json'),
        },
      },
```

This took my generation down from 2 minutes to 9 seconds.

Without the `tsconfig.json` file, the loaders can easily end up scanning an entire file system tree, which takes ages.

**Comment by shilman** [comment]
cc @strothj @clintandrewhall

**Comment by flxwu** [comment]
I've observed that babel-loader for me starts up a lot faster than atl - maybe it would make sense to change that in the sb documentation?

**Comment by maninacan** [comment]
I'm having a similar issue where starting SB pretty consistently takes about 10 minutes.  It will be going fine and then it hangs on some arbitrary file (it's a different file every time) and then it finishes out fine.  Meaning it only hangs once but it takes 10 minutes.  I'm on Storybook 5, but I'm not using TypeScript.  It was working fine with a few simple components but there are a couple of components I added that are now causing it to hang.  The components in question work fine in production.

**Comment by shilman** [comment]
@flxwu Mind submitting a PR for a docs change?

**Comment by shilman** [comment]
@maninacan Wanna jump on our discord? Maybe somebody from the community can help you debug it?

**Comment by felipenmoura** [comment]
This is really annoying and frustrating, and this is still happening to me!
I pinpointed it to this line:

```
@import './test.scss';
```

If I comment this line, it works fast and well. If I uncomment it, it stops at around 36%.

**Comment by shilman** [comment]
Interesting @felipenmoura . So the idea is that when you have multiple globs that resolve to the same story, some race condition is triggered that blows up the startup time?

**Comment by ndelangen** [comment]
I'm hoping that in the future, we don't rely on webpack's require.context at all anymore, and we resolve the globs (and can de-dup them) in node before starting webpack.

Sounds that might resolve that issue @shilman ?

**Comment by shilman** [comment]
@ndelangen that it would. it might also help with #12771 ?

**Comment by felipenmoura** [comment]
yup, exactly @shilman , this is not a critical problem, but it took me quite a while to figure it out :p 
I asked if this could become an issue because this is relatively intuitive to use it like that, and the problem was rather unexpected, you know.

**Comment by felipenmoura** [comment]
Oh my...are are facing this problem all over again!
But this time, it's happening only on a windows machine! Any ideas?!
